hotshot-query-service/src/data_source/fetching.rs

```diff
@@ -568,25 +622,57 @@ where
         self.fetcher.get(id.into()).await
     }
 
-    async fn get_block<ID>(&self, id: ID) -> Fetch<BlockQueryData<Types>>
+    async fn get_header<ID>(&self, id: ID) -> QueryResult<Fetch<Header<Types>>>
     where
         ID: Into<BlockId<Types>> + Send + Sync,
     {
-        self.fetcher.get(id.into()).await
+        if self.fetcher.leaf_only {
+            let mut tx = self.read().await.map_err(|err| QueryError::Error {
+                message: err.to_string(),
+            })?;
+            return tx.get_header(id.into()).await.map(Fetch::Ready);
+        }
+
+        let block: Fetch<BlockQueryData<Types>> = self.fetcher.get(id.into()).await;
```

Comment:
This could get the leaf instead of the block, which would not only be more efficient but would also work with the leaf-only version, so we could get rid of the conditional logic here

Reply:
Ohhh that's why :( We really should index leaves by hash and such as well, all the information is there...

In the meantime I guess this hack is ok

Reply:
I have added `HeaderQueryData` type which is just a wrapper but implements `Fetchable`. We can just subscribe to notifications for header request by block hash and payload hash. If we get a block or leaf then we take header from there and notify the subscribers

https://github.com/EspressoSystems/espresso-sequencer/pull/2529/commits/732b15af5adb8b9a59d704609fe17e054cbc03e1

Reply:
Beautiful!